fabric-registry-sync-v0/src/client/java/net/fabricmc/fabric/impl/client/registry/sync/RegistryRemovalChecker.java

```
@@ -0,0 +1,227 @@
+				states.add(property.getName());
+			}
+		}
+
+		JsonObject entries = new JsonObject();
+
+		for (Map.Entry<String, Map<String, Set<String>>> registry : entriesMap.entrySet()) {
+			JsonObject registryJson = new JsonObject();
+
+			for (Map.Entry<String, Set<String>> nsToEntries : registry.getValue().entrySet()) {
+				registryJson.add(nsToEntries.getKey(), Util.make(new JsonArray(), arr -> nsToEntries.getValue().forEach(arr::add)));
+			}
+
+			entries.add(registry.getKey(), registryJson);
+		}
+
+		JsonObject blockStates = new JsonObject();
+
+		for (Map.Entry<String, Map<String, Set<String>>> registry : blockStatesMap.entrySet()) {
+			JsonObject registryJson = new JsonObject();
+
+			for (Map.Entry<String, Set<String>> nsToEntries : registry.getValue().entrySet()) {
+				registryJson.add(nsToEntries.getKey(), Util.make(new JsonArray(), arr -> nsToEntries.getValue().forEach(arr::add)));
+			}
+
+			entries.add(registry.getKey(), registryJson);
+		}
+
+		json.add("entries", entries);
+		json.add("blockStates", blockStates);
+		return json;
+	}
+
+	@Nullable
+	public static RegistryRemovalChecker runCheck(Path jsonFile) {
+		if (DISABLED) {
+			LOGGER.info("Registry removal check was disabled via system property.");
+			return null;
+		}
+
+		try {
+			if (Files.exists(jsonFile)) {
+				// Do not run removal check if the file is missing
+				JsonObject json = JsonHelper.deserialize(Files.newBufferedReader(jsonFile, StandardCharsets.UTF_8));
+				return new RegistryRemovalChecker(json);
+			}
+		} catch (IOException | JsonParseException e) {
+			LOGGER.warn("Could not read {}, registry removal check disabled", FILE_NAME, e);
+		}
+
+		return null;
+	}
+
+	public static void write(Path jsonFile) {
+		if (DISABLED) return;
+
+		try {
+			JsonObject json = serializeRegistries();
+			Files.writeString(jsonFile, GSON.toJson(json), StandardCharsets.UTF_8);
+		} catch (IOException e) {
+			LOGGER.warn("Could not write {}", FILE_NAME, e);
```

Let it crash, something has gone badly wrong if this fails to write.